BackBLAST

```diff
@@ -474,19 +481,56 @@ function main() {
     printf "${SCRIPT_NAME}: pipeline to search for and visualize gene homologs across multiple genomes.\n"
     printf "Copyright Lee H. Bergstrand and Jackson M. Tsuji, Neufeld Research Group, 2019\n"
     printf "Version: ${VERSION}\n\n"
-    printf "Please specify a run mode for further usage instructions:\n"
+    printf "Please specify a run mode for the main workflow for further usage instructions:\n"
     printf "   1. setup: for setting up pre-run configuration files\n"
     printf "   2. run: to start a run using configuration files\n"
     printf "   3. auto: to skip setup and run the pipeline end-to-end with default values\n\n"
+    printf "Or run a specific step in the workflow manually:\n"
+    printf "   search: performs reciprocal BLASTP on given inputs\n"
+    printf "   remove_duplicates: removes duplicate hits from reciprocal BLASTP results\n"
+    printf "   create_blank_results: creates a blank output BLAST table for entries with no reciprocal BLASTP hits\n"
+    printf "   combine_tables: combines a set of input BLAST tables\n"
+    printf "   generate_heatmap: generates a heatmap from input BLAST and phylogeny data\n\n"
+    printf "Advanced parameters (use with care):\n"
+    printf "   -U utils_dir: Path to the BackBLAST utility script directory [default: ${UTILS_DIR}]\n\n"
 
     # Exit
     exit 0
-  fi    
+  fi
+
+  # Set defaults for options
+  local utils_dir
+  utils_dir="${UTILS_DIR}"
+
+  # Set options (help from https://wiki.bash-hackers.org/howto/getopts_tutorial; accessed March 8th, 2019)
+  OPTIND=1 # reset the OPTIND counter just in case
+  while getopts ":U:" opt; do
+    case ${opt} in
+      U)
+        utils_dir=${OPTARG}
+        ;;
+      \?)
+        echo "[ $(date -u) ]: ERROR: Invalid option: '-${OPTARG}'. Exiting..." >&2
+        exit 1
+        ;;
+      :)
+        echo "[ $(date -u) ]: ERROR: argument needed following '-${OPTARG}'. Exiting..." >&2
+        exit 1
+        ;;
+    esac
+  done
 
-  # Get user-supplied variables
+  # Set positional arguments
+  local original_arguments
+  original_arguments=${@} # save for reporting later
+  shift $((OPTIND - 1)) # shift to avoid flags when assigning positional arguments
   local run_mode
   run_mode=$1
 
+  # Load internal modules
+  # TODO: redesign the code so that this can be done at the top of the script, or run via CLI instead of loading as a module
+  source "${utils_dir}/generate_run_templates.sh" # the 'make_run_templates' function is called
```

Comment:
@LeeBergstrand I don't like having to do wait to source the script, but it's hard to avoid if UTILS_DIR can be modified by a user-provided flag. The only way I can think of to get around this would be to run the `generate_run_templates.sh` as a script through the CLI instead of sourcing it. I've already built the script to support CLI usage. Do you have any thoughts?

Reply:
Changed so that the CLI is accessed instead of loading the module.